### ISSUE TITLE
Perform autograd directly on Tensor, not Variable

### ISSUE THREAD

**Issue by adamlerer** [pytorch/pytorch#1384]
What does `Variable` do? Historically, it meant "wrapper for a tensor that can hold a graph and grad". But generally, a user only interacts with `Variable`s for which `requires_grad=False`, so it doesn't hold a graph or a grad. It's just a confusing wrapper to keep track of, with the additional confusion that you can't mix `Tensor` and `Variable` in torch ops (even though they're the same thing).

I think `Tensor` can directly have a `requires_grad` method and can optionally hold pointers to the grad and the graph. The advantages to the user are that they don't need to understand what a `Variable` is and they don't need to worry about calling `Variable` and `.data` at all the right places.

We can maybe do this while maintaining backwards compatibility with Variables (Variables can just be a thin wrapper around tensors that do nothing; and Tensor.data can just point at self).

**Comment by apaszke** [comment]
I think it's better to keep them separate - you have a tensor library supporting different kinds of devices + autograd on top, for the functions you want to differentiate. It'd be better to add auto-wrapping of Tensors (so `op(var, tensor)` is equivalent to `op(var, Variable(tensor))`).

**Comment by apaszke** [comment]
I've reopened #294.

**Comment by jcjohnson** [comment]
I'd agree for keeping them separate. One of the things I like about PyTorch in general is the separation of concerns between Tensors, Variables/autograd, and Modules; Tensor is just a generic ndarray that knows nothing about deep learning, or computational graphs, or gradients; it's just a numpy replacement that runs on GPU.

**Comment by adamlerer** [comment]
Yeah, automatic tensor promotion gets you most of the way there. The downside being you still have to deal with `.data` to convert back to tensors. But if torch ops accept tensors and variables interchangeably then you should rarely have to call `.data` either. There are some exceptions like dealing with single elements (python floats), or converting to numpy, etc.

I'm happy with #274 as a solution. @colesbury ?

**Comment by apaszke** [comment]
About dealing with single elements - I think it'd be the best to have a custom (possibly lazy) scalar type (like numpy does). It could use a 1-sized tensor to store it. This would allow us to e.g. avoid host sync when computing CUDA reductions, and would make Variables behave more like Tensors (e.g. you could no longer index them forever).
